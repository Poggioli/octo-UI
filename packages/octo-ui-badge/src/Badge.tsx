import { forwardRef } from "@octo-ui/core";
import { useMemo } from "react";
import { StyledBadge, StyledBadgeContainer } from "./styles";
import { BadgeProps } from "./types";

const Badge = forwardRef<typeof StyledBadge, BadgeProps>(
  (
    { count, maxCount = 99, showZero = false, dot = false, children, ...props },
    forwardedRef
  ) => {
    const showBadge = useMemo(() => showZero || count, [showZero, count]);

    const value = useMemo(
      () => (count > maxCount ? `${maxCount}+` : `${count}`),
      [count, maxCount]
    );

    return (
      <StyledBadgeContainer>
        {showBadge ? (
          <StyledBadge
            {...props}
            ref={forwardedRef}
            dot={dot}
          >
            {!dot ? value : null}
          </StyledBadge>
        ) : null}
        {children}
      </StyledBadgeContainer>
    );
  }
);
Badge.toString = () => `${StyledBadge.className}`;

export { StyledBadgeContainer, StyledBadge };
export { Badge };

