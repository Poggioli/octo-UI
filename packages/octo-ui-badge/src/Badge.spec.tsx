import { render } from "@testing-library/react";
import { Badge } from "./Badge";

describe("<Badge />", () => {
  it(`GIVEN a <Badge /> component
      WHEN render component
      SHOULD occurs render normaly`, () => {
    const rendered = render(
      <Badge count={10}>
        <button>Button badge</button>
      </Badge>
    );
    expect(rendered.container).toMatchSnapshot();
  });

  it(`GIVEN a <Badge /> component
      WHEN toString is called
      THEN return a string`, () => {
    expect(Badge.toString()).toBe("octo-UI-c-dqSAaa");
  });

  describe("dot props", () => {
    it(`GIVEN a <Badge /> component
        WHEN dot props is false
        THEN should show the count value`, () => {
      const rendered = render(
        <Badge count={10} dot={false}>
          <button>Button badge</button>
        </Badge>
      );

      const badgeCount = rendered.getByText("10");
      expect(badgeCount).toBeVisible();
      expect(rendered.container).toMatchSnapshot();
    });

    it(`GIVEN a <Badge /> component
        WHEN dot props is true
        THEN should NOT show the count value`, () => {
      const rendered = render(
        <Badge count={10} dot>
          <button>Button badge</button>
        </Badge>
      );

      const badgeCount = rendered.queryByText("10");
      expect(badgeCount).toBeNull();
      expect(rendered.container).toMatchSnapshot();
    });
  });

  describe("showZero props", () => {
    it(`GIVEN a <Badge /> component with count 0
        WHEN showZero props is false
        THEN should NOT show the count value`, () => {
      const rendered = render(
        <Badge count={0} showZero={false}>
          <button>Button badge</button>
        </Badge>
      );

      const badgeCount = rendered.queryByText("0");
      expect(badgeCount).toBeNull();
      expect(rendered.container).toMatchSnapshot();
    });

    it(`GIVEN a <Badge /> component with count 0
        AND showZero props is true
        THEN should show the count value`, () => {
      const rendered = render(
        <Badge count={0} showZero>
          <button>Button badge</button>
        </Badge>
      );

      const badgeCount = rendered.getByText("0");
      expect(badgeCount).toBeVisible();
      expect(rendered.container).toMatchSnapshot();
    });
  });

  describe("maxCount props", () => {
    it(`GIVEN a <Badge /> component
        WHEN maxCount is equal to 99 
        AND count props is less than 99
        THEN should show the count value`, () => {
      const rendered = render(
        <Badge count={3} showZero>
          <button>Button badge</button>
        </Badge>
      );

      const badgeCount = rendered.getByText("3");
      expect(badgeCount).toBeVisible();
      expect(rendered.container).toMatchSnapshot();
    });

    it(`GIVEN a <Badge /> component
      WHEN maxCount is equal to 99 
      AND count props is greater than 99
      THEN should render the "99+"`, () => {
      const rendered = render(
        <Badge count={100} showZero>
          <button>Button badge</button>
        </Badge>
      );

      const badgeCount = rendered.getByText("99+");
      expect(badgeCount).toBeVisible();
      expect(rendered.container).toMatchSnapshot();
    });
  });
});
