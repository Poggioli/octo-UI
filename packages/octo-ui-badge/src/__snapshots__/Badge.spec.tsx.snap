// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> GIVEN a <Badge /> component
      WHEN render component
      SHOULD occurs render normaly 1`] = `
<div>
  <span
    class="octo-UI-c-fONfOB"
  >
    <span
      class="octo-UI-c-dqSAaa octo-UI-c-dqSAaa-czsSGP-horizontalPosition-right octo-UI-c-dqSAaa-gCYbFg-verticalPosition-top octo-UI-c-dqSAaa-ixIDnb-color-unstyled octo-UI-c-dqSAaa-eatZjV-cv"
    >
      10
    </span>
    <button>
      Button badge
    </button>
  </span>
</div>
`;

exports[`<Badge /> dot props GIVEN a <Badge /> component
        WHEN dot props is false
        THEN should show the count value 1`] = `
<div>
  <span
    class="octo-UI-c-fONfOB"
  >
    <span
      class="octo-UI-c-dqSAaa octo-UI-c-dqSAaa-czsSGP-horizontalPosition-right octo-UI-c-dqSAaa-gCYbFg-verticalPosition-top octo-UI-c-dqSAaa-ixIDnb-color-unstyled octo-UI-c-dqSAaa-eatZjV-cv"
    >
      10
    </span>
    <button>
      Button badge
    </button>
  </span>
</div>
`;

exports[`<Badge /> dot props GIVEN a <Badge /> component
        WHEN dot props is true
        THEN should NOT show the count value 1`] = `
<div>
  <span
    class="octo-UI-c-fONfOB"
  >
    <span
      class="octo-UI-c-dqSAaa octo-UI-c-dqSAaa-czsSGP-horizontalPosition-right octo-UI-c-dqSAaa-gCYbFg-verticalPosition-top octo-UI-c-dqSAaa-ixIDnb-color-unstyled octo-UI-c-dqSAaa-cyzDYT-dot-true octo-UI-c-dqSAaa-eatZjV-cv"
    />
    <button>
      Button badge
    </button>
  </span>
</div>
`;

exports[`<Badge /> maxCount props GIVEN a <Badge /> component
        WHEN maxCount is equal to 99 
        AND count props is less than 99
        THEN should show the count value 1`] = `
<div>
  <span
    class="octo-UI-c-fONfOB"
  >
    <span
      class="octo-UI-c-dqSAaa octo-UI-c-dqSAaa-czsSGP-horizontalPosition-right octo-UI-c-dqSAaa-gCYbFg-verticalPosition-top octo-UI-c-dqSAaa-ixIDnb-color-unstyled octo-UI-c-dqSAaa-eatZjV-cv"
    >
      3
    </span>
    <button>
      Button badge
    </button>
  </span>
</div>
`;

exports[`<Badge /> maxCount props GIVEN a <Badge /> component
      WHEN maxCount is equal to 99 
      AND count props is greater than 99
      THEN should render the "99+" 1`] = `
<div>
  <span
    class="octo-UI-c-fONfOB"
  >
    <span
      class="octo-UI-c-dqSAaa octo-UI-c-dqSAaa-czsSGP-horizontalPosition-right octo-UI-c-dqSAaa-gCYbFg-verticalPosition-top octo-UI-c-dqSAaa-ixIDnb-color-unstyled octo-UI-c-dqSAaa-eatZjV-cv"
    >
      99+
    </span>
    <button>
      Button badge
    </button>
  </span>
</div>
`;

exports[`<Badge /> showZero props GIVEN a <Badge /> component with count 0
        AND showZero props is true
        THEN should show the count value 1`] = `
<div>
  <span
    class="octo-UI-c-fONfOB"
  >
    <span
      class="octo-UI-c-dqSAaa octo-UI-c-dqSAaa-czsSGP-horizontalPosition-right octo-UI-c-dqSAaa-gCYbFg-verticalPosition-top octo-UI-c-dqSAaa-ixIDnb-color-unstyled octo-UI-c-dqSAaa-eatZjV-cv"
    >
      0
    </span>
    <button>
      Button badge
    </button>
  </span>
</div>
`;

exports[`<Badge /> showZero props GIVEN a <Badge /> component with count 0
        WHEN showZero props is false
        THEN should NOT show the count value 1`] = `
<div>
  <span
    class="octo-UI-c-fONfOB"
  >
    <button>
      Button badge
    </button>
  </span>
</div>
`;
